typing/typeopt.ml

```diff
@@ -154,7 +154,7 @@ let bigarray_type_kind_and_layout env typ =
   | _ ->
       (Pbigarray_unknown, Pbigarray_unknown_layout)
 
-let value_kind env ty =
+let rec value_kind' env ~visited ty : Lambda.value_kind =
```

Comment:
I don't trust the use of a visited set here to make this function terminating. `scrape` is calling both `correct_levels` and `expand_head`, both of which can create fresh type nodes, and thus fresh type ids.

Since you are only recursing on product types, and recursive types without any sums in them are extremely rare, I think it would be better to just use some fuel to ensure termination in this case.

Also, please move this function within the body of `value_kind` and call it `loop` or similar. Leaving it as its own definition makes it seem like there are going to some uses of `value_kind'` outside of `value_kind`.

Reply:
I added some fuel. I think that visited is still valuable since there is really no point in unboxing recursive types.